### PR TITLE
automatically enable Horovod when multiple GPU IDs are specified

### PR DESCRIPTION
**Pull request by tkng**
This pull request is still work in progress, shutting down the process when training completed is not working correctly. (= We cannot exit the finished processes.)

## Motivation and Context

To train a large neural network, we'd like to use multiple GPUs. This pull request enables us to use multiple GPUs at the training phase of Blueoil.

## Description

Following changes are included:

- Introduced Horovod into the Blueoil Docker image
- If multiple GPUs are specified by `CUDA_VISIBLE_DEVICES`, automaticlly enables Horovod.
- Updated Horovod version to 0.16.4
- Updated OpenMPI version to 4.0.1

## How has this been tested?

There's no unit test for Horovod and it's a little bit hard to write. Hence I tested following situation.

- `CUDA_VISIBLE_DEVICES="0"`
- `CUDA_VISIBLE_DEVICES="1,2"`

In both cases, training was progressed expectedly.

## Screenshots (if appropriate):

None

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature / Optimization (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist:

- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.

### CODE REVIEW

**Comment by tkng** [comment]
I think now it's ready to merge, please review the PR.

**Comment by iizukak** [comment]
@nlpng 
Please review this PR.
I know you like Horovod 🙄

File: lmnet/executor/train.py [hunk 1]
```diff
@@ -112,7 +99,7 @@ def start_training(config):
         else:
             loss = model.loss(output, labels_placeholder)
         opt = model.optimizer(global_step)
```
Comment (yd8534976) [inline lmnet/executor/train.py]: Although I've never tried that. I heard that distributed training had a larger batch size. So the gradient is more accurate. But the update frequency is lower. To get a better performance, we usually need to enlarge learning rate to get a better performance. It should be something like `opt = tf.train.MomentumOptimizer(0.01 * hvd.size())`.

Comment (tkng) [reply lmnet/executor/train.py]: I has some experiments about that. The technique you mentioned is quite effective.

It would be a good idea to recommend to increase the learning rate in proportion to the batch size. However, automatically changing the learning rate which is set by a user, might confuse the user.

File: blueoil/blueoil_train.py [hunk 2]
```diff
@@ -20,17 +20,23 @@
...
 
 
 def run(blueoil_config_file, experiment_id):
     """Train from blueoil config."""
 
-    # Copy bueoil config yaml.
-    output_dir = os.environ.get('OUTPUT_DIR', 'saved')
-    experiment_dir = os.path.join(output_dir, experiment_id)
-    save_config_file(blueoil_config_file, experiment_dir)
+    if horovod_util.is_enabled():
```
Comment (yd8534976) [inline blueoil/blueoil_train.py]: I think this line is duplicated, you will call a `setup()` in below `is_rank0()`.

Comment (tkng) [reply blueoil/blueoil_train.py]: Thanks for the comment, there must be a room to improvement.

There're some cases we need to consider:

1. horovod is not installed and user doesn't use horovod
2. horovod is installed but user doesn't use horovod
3. horovod is not installed but use try to use horovod (should be an error in this case)
4. horovod is installed and user try to use horovod

Current code is not well considered to handle these 4 cases, I guess. Also, another problem is that `is_rank0` should not call `setup`.

Comment (tkng) [reply blueoil/blueoil_train.py]: I changed the code to exit immediately in the case of 3, since recover is impossible. I think now all cases are handled well.

Comment (yd8534976) [reply blueoil/blueoil_train.py]: LGTM.
